packages/trip-form/src/MetroModeSelector/utils.ts

```diff
@@ -0,0 +1,305 @@
+ */
+function useStateStorage<Type>(
+  name: string,
+  stateType: QueryParamConfig<Type>,
+  storeInQueryParam: boolean,
+  defaultState?: Type
+): [Type, (newValue: Type) => void] {
+  const [qpState, setQpState] = useQueryParam<Type>(name, stateType);
+  const [reactState, setReactState] = useState<Type>(defaultState);
+
+  if (!storeInQueryParam) {
+    return [reactState, setReactState];
+  }
+
+  if (qpState === undefined) {
+    return [defaultState, setQpState];
+  }
+  return [qpState, setQpState];
+}
+
+/**
+ * Grabs the activated modes and mode settings from the URL string
+ * when provided with all the necessary parameters. This allows access to the
+ * same parameters as useModeState in a context where React Hooks are not available,
+ * such as in Redux.
+ * @param searchString URL String
+ * @param modeButtons Mode button definitions (from config)
+ * @param modeSettingDefinitions Mode setting definitions (from config/defaults)
+ * @param initialState Initial state (from config)
+ * @returns Array of active TransportModes and all the mode settings
+ */
+export function getActivatedModesFromQueryParams(
+  searchString: string,
+  modeButtons: ModeButtonDefinition[],
+  modeSettingDefinitions: ModeSetting[],
+  initialState: InitialStateType
+): { activeModes: TransportMode[]; modeSettings: ModeSetting[] } {
+  const queryObject = new URLSearchParams(searchString);
+
+  const decodedQuery = decodeQueryParams(
+    { modeButtons: DelimitedArrayParam, modeSettings: ObjectParam },
+    {
+      modeButtons: queryObject.get("modeButtons"),
+      modeSettings: queryObject.get("modeSettings")
+    }
+  );
+
+  const enabledKeys =
+    decodedQuery.modeButtons || initialState.enabledModeButtons;
+  const activeButtons = filterModeDefitionsByKey(modeButtons, enabledKeys);
+
+  const modeSettingValues = {
+    // TODO: Do we want to keep these defaults with the definitions or only support it in initial state?
+    ...extractModeSettingDefaultsToObject(modeSettingDefinitions),
+    ...initialState.modeSettingValues,
+    ...decodedQuery.modeSettings
+  };
+
+  const modeSettingsWithValues = populateSettingsWithValues(
+    modeSettingDefinitions,
+    modeSettingValues
+  );
+  return {
+    activeModes: aggregateModes(activeButtons),
+    modeSettings: modeSettingsWithValues
+  };
+}
+
+/**
+ * This useModeState hook is designed to handle all the state relating to the mode selector.
+ * It consumes and provides all the needed data to make the mode selector work.
+ * @param buttonsFromConfig List of mode buttons to be displayed
+ * @param initialState Initial state object for mode button activation and mode setting defaults
+ * @param modeSettingDefinitions Definitions of all the mode settings available
+ * @param configuration Config to specify whether to store in url query param or react useState hook
+ * @returns Data to be used by mode selector and query generation
+ */
+export function useModeState(
```

Comment:
Are `useModeState` and `useStateStorage` intended to be used by implementers in their normal course of using the mode selector?